cmd/juju/common/controller.go

```diff
@@ -86,17 +91,23 @@ func WaitForAgentInitialisation(
 	}
 
 	ctx.Infof("Contacting Juju controller%s to verify accessibility...", addressInfo)
-	apiAttempts := 0
-	stop := make(chan struct{}, 1)
-	defer close(stop)
+
+	var apiAttempts int
 	err = retry.Call(retry.CallArgs{
 		Clock:    clock.WallClock,
 		Attempts: bootstrapReadyPollCount,
 		Delay:    bootstrapReadyPollDelay,
-		Stop:     stop,
+		Stop:     ctx.Context().Done(),
+		NotifyFunc: func(lastErr error, attempts int) {
+			apiAttempts = attempts
+		},
+		IsFatalError: func(err error) bool {
+			return errors.Is(err, &unknownError{}) ||
+				retry.IsRetryStopped(err) ||
+				errors.Is(err, stdcontext.Canceled) ||
+				errors.Is(err, stdcontext.DeadlineExceeded)
```

Comment:
This results in false positives for IsFatal for me

If tryAPI returns and error `*net.OpError, dial tcp x.x.x.x:17070: i/o timeout`, this matches stdcontext,DeadlineExceeded

However, in this case I believe we should retry

Reply:
> If tryAPI returns and error *net.OpError, dial tcp x.x.x.x:17070: i/o timeout, this matches stdcontext,DeadlineExceeded

This sounds like the point at which the net.OpError timeout error location should catch that error and return as a retryable error. A deadline exceed isn't retryable.

Reply:
@SimonRichardson this seems like a regression though. We're now seeing a new error when bootstrapping 3.2:
```
ERROR unable to contact api server after 0 attempts: unexpected error type: *net.OpError, dial tcp 172.31.1.196:17070: i/o timeout
```

Reply:
Just remove the deadline exceeded for now. We should revisit this in the future.

Reply:
That was my feeling. But I was holding off since I wasn't aware why it was included in the first place

Reply:
Do it 👍